pkg/controller/networkpolicy/clusternetworkpolicy.go

```diff
@@ -300,44 +302,80 @@ func (n *NetworkPolicyController) processClusterNetworkPolicy(cnp *crdv1alpha1.C
+				clusterPeers, perNSPeers := splitPeersByScope(cnpRule, direction)
+
+				// When a rule's NetworkPolicyPeer is empty, a cluster level rule should be created
+				// with an Antrea peer matching all addresses.
+				if len(clusterPeers) > 0 || len(perNSPeers) == 0 {
+					ruleAppliedTos := cnpRule.AppliedTo
+					// For ACNPs that have per-namespace rules, cluster-level rules will be created with appliedTo
+					// set as the spec appliedTo for each rule.
+					if appliedToPerRule && len(cnp.Spec.AppliedTo) > 0 {
+						ruleAppliedTos = cnp.Spec.AppliedTo
+					}
+					ruleATGNames := n.processClusterAppliedTo(ruleAppliedTos, atgNamesSet)
+					klog.V(4).Infof("Adding a new cluster-level rule with appliedTos %v for %s", ruleATGNames, cnp.Name)
+					addRule(n.toAntreaPeerForCRD(clusterPeers, cnp, direction, namedPortExists), services, direction, ruleATGNames)
 				}
-				ruleATGNames := n.processClusterAppliedTo(ruleAppliedTos, atgNamesSet)
-				klog.V(4).Infof("Adding a new cluster-level rule with appliedTos %v for %s", ruleATGNames, cnp.Name)
-				addRule(n.toAntreaPeerForCRD(clusterPeers, cnp, direction, namedPortExists), direction, ruleATGNames)
-			}
-			if len(perNSPeers) > 0 {
-				if len(cnp.Spec.AppliedTo) > 0 {
-					// Create a rule for each affected Namespace of appliedTo at spec level
-					for i := range clusterAppliedToAffectedNS {
-						klog.V(4).Infof("Adding a new per-namespace rule with appliedTo %v for rule %d of %s", clusterAppliedToAffectedNS[i], idx, cnp.Name)
-						addRule(n.toNamespacedPeerForCRD(perNSPeers, clusterAppliedToAffectedNS[i]), direction, []string{atgForNamespace[i]})
+				if len(perNSPeers) > 0 {
+					if len(cnp.Spec.AppliedTo) > 0 {
+						// Create a rule for each affected Namespace of appliedTo at spec level
+						for i := range clusterAppliedToAffectedNS {
+							klog.V(4).Infof("Adding a new per-namespace rule with appliedTo %v for rule %d of %s", clusterAppliedToAffectedNS[i], idx, cnp.Name)
+							addRule(n.toNamespacedPeerForCRD(perNSPeers, clusterAppliedToAffectedNS[i]), services, direction, []string{atgForNamespace[i]})
+						}
+					} else {
+						// Create a rule for each affected Namespace of appliedTo at rule level
+						for _, at := range cnpRule.AppliedTo {
+							affectedNS, selectors := n.getAffectedNamespacesForAppliedTo(at)
+							affectedNamespaceSelectors = append(affectedNamespaceSelectors, selectors...)
+							for _, ns := range affectedNS {
+								atg := n.createAppliedToGroup(ns, at.PodSelector, nil, at.ExternalEntitySelector)
+								atgNamesSet.Insert(atg)
+								klog.V(4).Infof("Adding a new per-namespace rule with appliedTo %v for rule %d of %s", atg, idx, cnp.Name)
+								addRule(n.toNamespacedPeerForCRD(perNSPeers, ns), services, direction, []string{atg})
+							}
+						}
+					}
+				}
+				klog.Infof("Checking cluster or perNSPeers: cluster:%d, perNS: %d", len(clusterPeers), len(perNSPeers))
+			} else {
```

Comment:
pure nit: this has become a gigantic function. Consider breaking it if possible.

Reply:
Done.